src/memray/_memray/tracking_api.cpp

```diff
@@ -284,6 +321,101 @@ std::unique_ptr<Tracker> Tracker::d_instance_owner;
 std::atomic<Tracker*> Tracker::d_instance = nullptr;
 MEMRAY_FAST_TLS thread_local size_t NativeTrace::MAX_SIZE{128};
 
+std::vector<PythonStackTracker::LazilyEmittedFrame>
+PythonStackTracker::pythonFrameToStack(PyFrameObject* current_frame)
+{
+    std::vector<LazilyEmittedFrame> stack;
+
+    while (current_frame) {
+        PyCodeObject* code = compat::frameGetCode(current_frame);
+
+        const char* function = PyUnicode_AsUTF8(code->co_name);
+        if (function == nullptr) {
+            Py_DECREF(code);
+            return {};
+        }
+
+        const char* filename = PyUnicode_AsUTF8(code->co_filename);
+        if (filename == nullptr) {
+            Py_DECREF(code);
+            return {};
+        }
+
+        int lineno = PyFrame_GetLineNumber(current_frame);
+        stack.push_back({current_frame, {function, filename, lineno}, false});
+        current_frame = compat::frameGetBack(current_frame);
+    }
+
+    return stack;
+}
+
+void
+PythonStackTracker::recordAllStacks()
+{
+    assert(PyGILState_Check());
+
+    // Record the current Python stack of every thread
+    std::unordered_map<PyThreadState*, std::vector<LazilyEmittedFrame>> stack_by_thread;
+    for (PyThreadState* tstate =
+                 PyInterpreterState_ThreadHead(compat::threadStateGetInterpreter(PyThreadState_Get()));
+         tstate != nullptr;
+         tstate = PyThreadState_Next(tstate))
+    {
+        PyFrameObject* frame = compat::threadStateGetFrame(tstate);
+        if (frame) {
+            stack_by_thread[tstate] = pythonFrameToStack(frame);
+            if (PyErr_Occurred()) {
+                throw std::runtime_error("Failed to capture a thread's Python stack");
+            }
+        }
```

Comment:
Sorry, I am allergic to indentation

```suggestion
        if (!frame) {
            continue;
        }            
        stack_by_thread[tstate] = pythonFrameToStack(frame);
        if (PyErr_Occurred()) {
            throw std::runtime_error("Failed to capture a thread's Python stack");
        }
```